File: eTouristAgencyAPI/eTouristAgencyAPI.Services/Database/Models/Reservation.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace eTouristAgencyAPI.Services.Database.Models;

[Table("Reservation")]
public partial class Reservation
{
    [Key]
    public Guid Id { get; set; }

    public Guid UserId { get; set; }

    public DateTime CreatedOn { get; set; }

    public DateTime ModifiedOn { get; set; }

    public DateTime? CancellationDate { get; set; }

    [Column(TypeName = "decimal(10, 2)")]
    public decimal TotalCost { get; set; }

    public Guid? OfferDiscountId { get; set; }

    public Guid ReservationStatusId { get; set; }

    public Guid ModifiedBy { get; set; }

    public Guid CreatedBy { get; set; }

    public Guid RoomId { get; set; }

    public long ReservationNo { get; set; }

    [Column(TypeName = "decimal(15, 2)")]
    public decimal PaidAmount { get; set; }

    public string? Note { get; set; }

    [ForeignKey("CreatedBy")]
    [InverseProperty("ReservationCreatedByNavigations")]
    public virtual User CreatedByNavigation { get; set; } = null!;

    [ForeignKey("ModifiedBy")]
    [InverseProperty("ReservationModifiedByNavigations")]
    public virtual User ModifiedByNavigation { get; set; } = null!;

    [ForeignKey("OfferDiscountId")]
    [InverseProperty("Reservations")]
    public virtual OfferDiscount? OfferDiscount { get; set; }

    [InverseProperty("Reservation")]
    public virtual ICollection<Passenger> Passengers { get; set; } = new List<Passenger>();

    [InverseProperty("Reservation")]
    public virtual ICollection<ReservationPayment> ReservationPayments { get; set; } = new List<ReservationPayment>();

    [InverseProperty("IdNavigation")]
    public virtual ReservationReview? ReservationReview { get; set; }

    [ForeignKey("ReservationStatusId")]
    [InverseProperty("Reservations")]
    public virtual EntityCodeValue ReservationStatus { get; set; } = null!;

    [ForeignKey("RoomId")]
    [InverseProperty("Reservations")]
    public virtual Room Room { get; set; } = null!;

    [ForeignKey("UserId")]
    [InverseProperty("ReservationUsers")]
    public virtual User User { get; set; } = null!;

    [NotMapped]
    public bool IsEditable { get; set; }

    [NotMapped]
    public bool IsReviewable { get; set; }
}

File: Frontend/etouristagency_mobile/lib/models/reservation/reservation.dart
import 'package:etouristagency_mobile/models/entity_code_value/entity_code_value.dart';
import 'package:etouristagency_mobile/models/offer_discount/offer_discount.dart';
import 'package:etouristagency_mobile/models/passenger/passenger.dart';
import 'package:etouristagency_mobile/models/reservation/reservation_payment.dart';
import 'package:etouristagency_mobile/models/room/room.dart';
import 'package:etouristagency_mobile/models/user/user.dart';

class Reservation {
  String? id;
  String? userId;
  DateTime? createdOn;
  DateTime? modifiedOn;
  double? paidAmount;
  DateTime? cancellationDate;
  int? reservationNo;
  double? totalCost;
  String? offerDiscountId;
  String? reservationStatusId;
  String? roomId;
  Room? room;
  OfferDiscount? offerDiscount;
  List<Passenger>? passengers;
  EntityCodeValue? reservationStatus;
  List<ReservationPayment>? reservationPayments;
  User? user;
  String? note;

  Reservation(
    this.id,
    this.userId,
    this.createdOn,
    this.modifiedOn,
    this.paidAmount,
    this.cancellationDate,
    this.reservationNo,
    this.totalCost,
    this.offerDiscountId,
    this.reservationStatusId,
    this.roomId,
    this.room,
    this.offerDiscount,
    this.passengers,
    this.reservationStatus,
    this.reservationPayments,
    this.user,
    this.note
  );

  factory Reservation.fromJson(Map<String, dynamic> json) {
    return Reservation(
      json["id"],
      json["userId"],
      json["createdOn"] != null ? DateTime.parse(json["createdOn"]) : null,
      json["modifiedOn"] != null ? DateTime.parse(json["modifiedOn"]) : null,
      json["paidAmount"] != null
          ? (json["paidAmount"] as num).toDouble()
          : null,
      json["cancellationDate"] != null
          ? DateTime.parse(json["cancellationDate"])
          : null,
      json["reservationNo"],
      json["totalCost"] != null ? (json["totalCost"] as num).toDouble() : null,
      json["offerDiscountId"],
      json["reservationStatusId"],
      json["roomId"],
      json["room"] != null ? Room.fromJson(json["room"]) : null,
      json["offerDiscount"] != null
          ? OfferDiscount.fromJson(json["offerDiscount"])
          : null,
      json["passengers"] != null
          ? (json["passengers"] as List)
                .map((e) => Passenger.fromJson(e))
                .toList()
          : null,
      json["reservationStatus"] != null
          ? EntityCodeValue.fromJson(json["reservationStatus"])
          : null,
      json["reservationPayments"] != null ? (json["reservationPayments"] as List).map((e)=> ReservationPayment.fromJson(e)).toList() : null,
      json["user"] != null ? User.fromJson(json["user"]) : null,
      json["note"]
    );
  }
}
